File: src/main/java/sit/bookregistry/service/UserService.java
package sit.bookregistry.service;

import sit.bookregistry.model.Book;
import sit.bookregistry.model.User;
import sit.bookregistry.repository.BookRepository;
import sit.bookregistry.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.security.InvalidParameterException;
import java.util.List;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final BookRepository bookRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, BookRepository bookRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.bookRepository = bookRepository;
        this.passwordEncoder = passwordEncoder;
    }

    public void register(String email, String password, String repeatedPassword) {
        if (!password.equals(repeatedPassword))
            throw new IllegalArgumentException("passwords must match!");
        try {
            userRepository.save(new User(email, passwordEncoder.encode(password), false));
        }catch (Exception ex){
            throw new RuntimeException("User with that email already exists!");
        }
    }

  
    public void borrowBook(Long bookId, String email) {
        User user = getUserByUsername(email);
        Book book = bookRepository.findById(bookId).orElseThrow(() -> new RuntimeException("Book does not exist"));
        if (book.getQuantity() == 0) throw new RuntimeException("No books left");
        user.getBooks().add(book);
        userRepository.save(user);
        book.setQuantity(book.getQuantity() - 1);
        bookRepository.save(book);
    }

    public List<Book> getUserBooks(String email) {
        return getUserByUsername(email).getBooks();
    }

    private User getUserByUsername(String username) {
        return userRepository.findUserByUsername(username)
                .orElseThrow(() -> new RuntimeException("User does not exist"));
    }

    
    public void returnBook(Long bookId, String email) {
        User user = getUserByUsername(email);
        Book book = user.getBooks()
                .stream()
                .filter(userBook -> userBook.getId().equals(bookId))
                .findAny()
                .orElseThrow(() -> new RuntimeException("User does not own this book"));
        user.getBooks().remove(book);
        userRepository.save(user);
        book.setQuantity(book.getQuantity() + 1);
        bookRepository.save(book);
    }

    public boolean login(String email, String password) {
        User user = getUserByUsername(email);
        if (!passwordEncoder.matches(password, user.getPassword()))
            throw new InvalidParameterException("Invalid credentials");

        return user.isAdmin();
    }
}

File: book-registry-frontend/src/app/book-registry/service/user/user.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Book} from "../../shared/book";
import {UserBook} from "../../shared/user-book";

const API_ENDPOINT = 'user/';

@Injectable({
  providedIn: 'root'
})
export class UserService {

  constructor(private http: HttpClient) {
  }

  login(user): Observable<boolean> {
    return this.http.post<boolean>(API_ENDPOINT + 'login', user);
  }

  register(user): Observable<void> {
    return this.http.post<void>(API_ENDPOINT + 'register', user);
  }

  borrowBook(value: UserBook): Observable<void> {
    return this.http.post<void>(API_ENDPOINT + 'borrow-book', value);
  }

  returnBook(value): Observable<void> {
    return this.http.post<void>(API_ENDPOINT + 'return-book', value);
  }

  getUserBooks(username: string): Observable<Book[]> {
    return this.http.get<Book[]>(API_ENDPOINT + 'books?username=' + username);
  }
}
